fggs/indices.py

```diff
@@ -0,0 +1,801 @@
+                               self.pembeds, {})[0].clone()
+        virtual = self.physical.new_full(self.size(), self.default)
+        # TODO: allow pembeds_fv <= vembeds_fv by repeating self.physical?
+        project(virtual, self.pembeds, self.vembeds, {})[0].copy_(self.physical)
+        return virtual
+
+    def dim_to_dense(self, dim) -> EmbeddedTensor:
+        """Expand to an equivalent EmbeddedTensor but make sure that the given
+           dimension is dense and independent of the other dimensions."""
+        vembeds = list(self.vembeds)
+        e_dense = vembeds.pop(dim)
+        rename : Rename = {}
+        vembeds = list(e.freshen(rename) for e in vembeds)
+        if isinstance(e_dense, EmbeddingVar) and e_dense not in rename:
+            return self
+        fv  = tuple(rename.keys())
+        fv0 = tuple(rename.values())
+        k = EmbeddingVar(e_dense.numel())
+        vembeds.insert(dim, k)
+        pembeds = fv0 + (k,)
+        default = self.default
+        physical = self.physical.new_full(Size(e.numel() for e in pembeds), default)
+        project(physical, self.pembeds, fv + (e_dense,), {})[0].copy_(self.physical)
+        return EmbeddedTensor(physical, pembeds, vembeds, default)
+
+    def permute(self, dims: Sequence[int]) -> EmbeddedTensor:
+        assert(len(dims) == len(self.vembeds))
+        assert(frozenset(dims) == frozenset(range(0,len(dims))))
+        return EmbeddedTensor(self.physical,
+                              self.pembeds,
+                              tuple(self.vembeds[i] for i in dims),
+                              self.default)
+
+    def __iter__(self) -> Iterator[EmbeddedTensor]:
+        self = self.dim_to_dense(0)
+        vembeds = list(self.vembeds)
+        k = vembeds.pop(0)
+        assert(isinstance(k, EmbeddingVar))
+        pembeds = list(self.pembeds)
+        i = pembeds.index(k)
+        pembeds.pop(i)
+        perm = (i, *range(0, i), *range(i+1, len(self.pembeds)))
+        for p in self.physical.permute(perm):
+            yield EmbeddedTensor(p, pembeds, vembeds, self.default)
+
+    def relu_without_nan_(self) -> EmbeddedTensor:
+        self.default = max(0, self.default) # max(0, math.nan) == 0 != max(math.nan, 0)
+        self.physical.relu_().nan_to_num_(nan=0., posinf=inf)
+        return self
+
+    def __imul__(self, other: NumberType) -> EmbeddedTensor:
+        self.default  *= other
+        self.physical *= other
+        return self
+
+    def __itruediv__(self, other: NumberType) -> EmbeddedTensor:
+        self.default  /= other
+        self.physical /= other
+        return self
+
+    def exp(self) -> EmbeddedTensor:
+        return EmbeddedTensor(self.physical.exp(), self.pembeds, self.vembeds, exp(self.default))
+
+    def logical_not(self) -> EmbeddedTensor:
+        return EmbeddedTensor(~self.physical, self.pembeds, self.vembeds, not self.default)
+
+    def log_softmax(self, dim) -> EmbeddedTensor:
```

Comment:
sum_product.log_softmax has some extra checks for infinite inputs which are important. I think this drops those checks, right?

Reply:
Just to clarify, the comment "make all the nonzero elements 1 and the zero elements 0" in sum_product.log_softmax means that the final result (due to log) is actually either 0=log(1) or -inf=log(0), right?